### ISSUE TITLE
Native GTD "Waiting For" report

### ISSUE THREAD

**Issue by JonasDralle** [GothenburgBitFactory/taskwarrior#2221]
Hi. I'm doing GTD with TaskWarrior and I tag items with a `waitfor` tag in order to "wait for" someone else to do this task. It's quite handy and a central part of doing GTD.

Can someone add a more elegant way of dong this? Having a magical tag requires all reports fo filter waitfor-Tasks out. Having waitfor-Tasks is very handy since one can have dependencies on them.
e. g. "I can't do the report because bob hasn't sent me the figures yet".

This is something which GTD users would enjoy and Non-GTD users could ignore easily.

**Comment by Xerus2000** [comment]
Can't you simply add a wait date sometime in the future and add a tag for the custom view? 
https://taskwarrior.org/docs/terminology.html#waiting

**Comment by Xerus2000** [comment]
I think what you are looking for is wait:someday - see https://taskwarrior.org/docs/named_dates.html

**Comment by JavaZauber** [comment]
> I think what you are looking for is wait:someday - see https://taskwarrior.org/docs/named_dates.html#calculated

No, having an item on the 'Waiting for' list in contrast to the 'Someday/Maybe' list are conceptionally different things in the world of GTD.

![excerpt1](https://user-images.githubusercontent.com/6533815/78905378-2dbc3b00-7a7e-11ea-836b-e480e1d33567.png)
versus
![excerpt2](https://user-images.githubusercontent.com/6533815/78905045-aff82f80-7a7d-11ea-9012-84243b5ebb25.png)
[Full Diagram](https://rufuspollock.com/2019/05/27/getting-things-done/gtd-clarify-flow.png) ; [Overview of GTD](https://hamberg.no/gtd/)

> add a tag for the custom view

I already have a custom report, as stated in the issue description.
The formatting is like this:

    ID  Created    Description                                                                Urg 
    123 2020-01-01 [Marko] Send files regarding  trampoline                                   0.00
    321 2020-01-01 [Christine] Return call                                                    0.00

**Comment by Xerus2000** [comment]
you are totally right, but I think you should have a `wait:someday` attached to both waiting for and someday tasks, with an extra `+waiting` tag for the tasks you are waiting on - that way both don't show up in the queries and you can still have your custom report.

**Comment by apkawel** [comment]
I might be missing something, @JavaZauber, but I think you could just create a context that removes tasks that have your tag. That way you can simply activate that context and then all the reports will respect the context's constraints when you run them.

Try this in your `.taskrc` file:

`context.notwaiting=-waitfor`

Then switch to the context and run the report:
```
$ task context notwaiting
$ task next
```
… or any other report. That should do the trick.

Of course you can add or subtract other tags as needed; this is just a minimal working example.

**Comment by linuxluser** [comment]
The GTD "Waiting For" list is not the same as the TW concept of "wait" or "waiting".

In GTD, the underlying concept is *delegating tasks to others*.

Check out the original GTD flowchart (I've annotated the important part for our discussion):

<img width="775" height="1013" alt="Image" src="https://github.com/user-attachments/assets/5db139fd-d413-4181-a3f4-7284c7e6d293" />

The "Waiting For" list in GTD is really about tracking what you've delegated to *other people* to do. To implement this in TW, you'd need the concept of a task being "owned" by someone else other than the current user. It doesn't have to be complicated. It's really just that you don't want tasks owned by someone else to show up in your own task reports. But you might want to associate a delegated task with a project.

I don't want to tell anybody how to implement this but my own concept of it is that it could just be a new task state: `delegated`. A delegated task is one that you want to track and and associate with projects or other tags, but that you don't want cluttering your normal reports.

For example, a manager who is tracking multiple projects and potentially hundreds of tasks delegated to others would find something like this extremely valuable. And the TW report could just be `task delegated`. This would, in essence, be the GTD "Waiting For" list but you could avoid the keyword "wait" so that existing TW concepts would stay in-tact.

**Comment by ChargingBulle** [comment]
I agree with what you say.
I usually look at my list tagged `+waitfor` during weekly review (I also use `tasksh` for this!).
`status:delegated` makes sense to me

The `wait:` field does something different because it hides the item until the time is reached.

**Comment by linuxluser** [comment]
Right. `waiting` in TW is more about the time aspect of a task. It's GTD equivalent would actually be the "Tickler File" rather than the "Waiting For" list. The Tickler File is a system to remind yourself at a future date of something that needs to be (potentially) put into your active lists. This is exactly what `wait:2026-12-25` is made for!

So, I agree that there just isn't a functionality in TW at the moment that can implement the GTD "Waiting For" list cleanly.

I am maintining a list outside of TW for this. I am already maintaining a projects list outside of TW as well (because TW doesn\'t really support that very cleanly either) so I might as well maintain the Waiting For list as well.

Another option would be to create a UDA (say, `owner`, e.g. `owner:Steve`) and updating the reports you use to filter out tasks with an owner. The problem with this is that I use a mobile client and the sync server and the mobile client won't filter those out. So it wouldn't be a consistent experience across clients.

A final option would be to use a separate TW config and database to house tasks owned by others. If I were needing to track 100s of tasks for others, this is what I'd do. But in my case, my "Waiting For" list is usually less than 15 items so it's not a huge deal to track it in a simple text file.

**Comment by linuxluser** [comment]
This has been racking my brain for a couple of days now. Because this ought to be easy to implement as a UDA but if you do that, you have to update each and every one of the reports as well, including when you add new reports later. Which is fine but there should be a more elegant way.

There are a few approaches I've thought of.

**1. Make a new new property `Owner` of tasks.**

This would require code changes that default to only showing tasks owned by current user (empty value) unless queried for `owner:<specific name>`. This isn't too different than just having a tag of someone's name but internally, TW would know not to show tasks owned by others by default. And there would be a new report called `delegated` to show all the delegated tasks (tasks owned by others). To facilitate that, we'd probably need a new virtual tag `DELEGATED` so that the code could easily check and know what to do and so that people could add `+DELEGATED` to their filters and reports whenever they want to see tasks owned by others.

**2. Create a virtual tag `HIDDEN` for tasks and allow a UDA to set this by default.**

This would extend UDA's a bit and allow people to specify that, by default, they'd like tasks that have their UDA to not be visible in reports. The default, of course, is that UDA are always visible but this would allow inverting this to say that a UDA is invisible (hidden) by default unless specifically requested to be visible in a filter/report.

When you define your UDA, you could set the `visibility` property. So, something like `uda.myuda1.visibility=hidden`. This is parallel to how HTML elements work.

---

I think \#2 is more complicated but would open things up for more interesting kinds of UDAs for people so that it wouldn't just be waiting tasks or recurring tasks that get the special treatment of being hidden by default. This would approach would generalize that concept and give people more options. But maybe it is a bit too technical?

Option \#1 would be the most intuitive to people since it's simply a new, first-class feature. It could have its own documentation page to show how you can now easily track tasks for other people and not just yourself. And it could be built upon in the future to add useful things to it that I haven't thought of yet.

Currently I'm leaning to option \#1. But I'd really like to get a developer's perspective. I've looked at the code a little and I think I could help implement some of this.

**Comment by ChargingBulle** [comment]
> I am already maintaining a projects list outside of TW as well (because TW doesn't really support that very cleanly either) 

very true!
TW hides projets with zero open items but it's exactly those "active projects with no next actions" which require your attentionin the GTD world. I have my project list inside the Horizonts of Focus LaTeX document of mine.
Ngl it's quite easy to solve this by just having a .task/projects.txt which you maintain with an editor with every line a string matching a project.

My far bigger concern which has nothing to do with this ticket is that TaskWarrior V3 is really really slow.
especially batch operations have been instant in the past (e. g. `task 10-20 mod project:foo`) but now since the new database has been rebuild you can really watch it process one task at a time at gruesome slow speeds

**Comment by ChargingBulle** [comment]
I have a report which only shows `+waitfor` and for every other report I do `-waitfor`. That has been my solution since I created this ticket in 2019. I just write the name of the person I am waiting for in the title of the ticket e. g. 
```
task waitfor
4  [Dad] Return the lawn mower
31 [Sven] give me back the 20 € he borrowed
```

**Comment by linuxluser** [comment]
One more idea I've had is that maybe a way to implement this would be to have TW hold an internal (virtual) "task" that represents outside factors (call it *EXTERNAL* and assign it a static UUID) and real tasks can be set to depend on that internal "task". So you'd set it via `depends:external` (alias of `depends:others`, `depends:someone`, etc).

This might be more feasible to implement as it should really only require that TW have a static task be compiled into the program and referencable by a keyword (`external`) rather than an ID or UUID. And with this setup, maybe a shorthand could be created with the special tag `waitingfor` or something that makes that task blocked on the internal task.

So `task add '[Karen] Mail the checks' +waitingfor` would make a new task that's blocked by the `EXTERNAL` built-in task. Might even make it so it's got its own color and such in themes.
